src/server/debug.rs

```diff
@@ -1165,6 +1276,86 @@ fn set_region_tombstone(db: &DB, store_id: u64, region: Region, wb: &WriteBatch)
     Ok(())
 }
 
+fn divide_db(db: &DB, parts: usize) -> ::raftstore::Result<Vec<Vec<u8>>> {
+    let mut fake_region = Region::default();
+    fake_region.set_start_key(vec![]);
+    fake_region.set_end_key(vec![]);
+    fake_region.mut_peers().push(Peer::default());
+    let get_cf_size =
+        |cf: &str| raftstore_util::get_region_approximate_size_cf(db, cf, &fake_region);
+
+    let default_cf_size = box_try!(get_cf_size(CF_DEFAULT));
+    let write_cf_size = box_try!(get_cf_size(CF_WRITE));
+
+    let middle_by_cf = if default_cf_size >= write_cf_size {
+        CF_DEFAULT
+    } else {
+        CF_WRITE
+    };
+
+    divide_db_cf(db, parts, middle_by_cf)
+}
+
+fn divide_db_cf(db: &DB, parts: usize, cf: &str) -> ::raftstore::Result<Vec<Vec<u8>>> {
+    let cf = get_cf_handle(db, cf)?;
+    let start = keys::data_key(b"");
+    let end = keys::data_end_key(b"");
+    let range = Range::new(&start, &end);
+    let collection = db.get_properties_of_tables_in_range(cf, &[range])?;
+
+    let mut keys = Vec::new();
+    let mut found_keys_count = 0;
+    for (_, v) in &*collection {
+        let props = RangeProperties::decode(v.user_collected_properties())?;
+        keys.extend(
+            props
+                .offsets
+                .range::<[u8], _>((Excluded(start.as_slice()), Excluded(end.as_slice())))
+                .filter(|_| {
+                    found_keys_count += 1;
+                    found_keys_count % 100 == 0
+                })
+                .map(|(k, _)| k.to_owned()),
+        );
+    }
+
+    println!(
+        "({} points found, {} points selected for dividing)",
+        found_keys_count,
+        keys.len()
+    );
+
+    if keys.is_empty() {
+        return Ok(vec![]);
+    }
+
+    if keys.len() > 20000 {
+        let len = keys.len();
+        keys = keys.into_iter().step_by(len / 10000).collect();
+    }
+
+    keys.sort();
+
+    Ok(sample(keys, parts - 1))
+}
+
+fn sample<T>(data: Vec<T>, samples: usize) -> Vec<T> {
+    if samples == 0 {
+        return vec![];
+    }
+
+    if samples >= data.len() {
+        return data;
+    }
+
+    let len = data.len();
+    data.into_iter()
+        .enumerate()
+        .filter(|(i, _)| i + 1 != len && i * (samples + 1) / len != (i + 1) * (samples + 1) / len)
```

Comment:
It takes me some time to understand it. 😸

Reply:
This is used to find all dividing points and I have no idea how to simplify the calculation..